### PR TITLE
feat(sn_node): let elder store chunks as well

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: sn_node/src/node/messaging/client_msgs.rs [hunk 1]
```diff
@@ -177,8 +178,58 @@ impl MyNode {
...
-            DataCmd::StoreChunk(chunk) => Ok(ReplicatedData::Chunk(chunk)),
+            DataCmd::StoreChunk(chunk) => {
+                let data = ReplicatedData::Chunk(chunk);
+                info!("Try to store the data {:?} in elder locally", data.name());
+                // TODO: check the storage level first.
+                //       If over 50% full, remove old entries to free up space for new.
+                let section_pk = PublicKey::Bls(context.network_knowledge.section_key());
+                let node_keypair = Keypair::Ed25519(context.keypair.clone());
+                let store_result = context
+                    .data_storage
```
Comment (joshuef) [inline sn_node/src/node/messaging/client_msgs.rs:189]: Again, i wouldn't treat the elder as special here. If it's a holder, we request it (of ourselves)

**Review by joshuef — CHANGES_REQUESTED** [review]
I think we're also missing a change to the `get_data_holders` functions, as they don't include elders at the moment?

File: sn_node/src/node/data/records.rs [hunk 2]
```diff
@@ -219,15 +221,43 @@ impl MyNode {
         }
     }
 
+    /// Being an elder, try to fetch the data locally first. If not having, fetch from adults.
     /// Find target adult, sends a bidi msg, awaiting response, and then sends this on to the client
-    pub(crate) async fn read_data_from_adult_and_respond_to_client(
+    pub(crate) async fn read_data_and_respond_to_client(
```
[the remark this answers is not in the document]
Comment (maqi) [reply sn_node/src/node/data/records.rs:226]: yes, I have changed the function to query self first, then the adults. hence the function name is changed.

File: sn_node/src/node/messaging/client_msgs.rs [hunk 3]
```diff
@@ -177,8 +178,58 @@ impl MyNode {
 
         // extract the data from the request
         let data_result = match cmd.clone() {
-            // These reads/writes are for adult nodes...
-            DataCmd::StoreChunk(chunk) => Ok(ReplicatedData::Chunk(chunk)),
+            DataCmd::StoreChunk(chunk) => {
+                let data = ReplicatedData::Chunk(chunk);
+                info!("Try to store the data {:?} in elder locally", data.name());
```
[the remark this answers is not in the document]
Comment (maqi) [reply sn_node/src/node/messaging/client_msgs.rs:183]: This is for `storing` flow ?

File: sn_node/src/node/messaging/client_msgs.rs (hunk 1 above)
Comment (maqi) [reply sn_node/src/node/messaging/client_msgs.rs:189]: You mean treat self as an adult ?
that will make existing flow become tricky and un-necessary for some part.
such as the `record storage level` , `correlation id`, `fit into the specific message type between Adult and Elder` , etc.

**Comment by maqi** [comment]
> I think we're also missing a change to the `get_data_holders` functions, as they don't include elders at the moment?

I think we don't need to do so, as the elder can check its local storage directly.
In case it doesn't have, with high chance other elders won't have either, hence can contact adults directly.
Even other elders have the copy, the client shall contact them directly already, hence no need for this elder to query other elders.

**Comment by maqi** [comment]
PR updated to make elder storing data as normal adults

**Comment by maqi** [comment]
work included by the PR https://github.com/maidsafe/safe_network/pull/1926

**Pull request closed without merging**